### ISSUE TITLE
Merging climate data & weights

### ISSUE THREAD

**Issue by saraorofino** [tcarleton/climate-aggregate#5]
Hi @tcarleton I'm working on this next step in the pipeline and wanted to clarify a few things: 

- Inputs: so far inputs I have are climate variable, year(s), data source, and something related to weights I haven't figured out yet
     - Are we doing a single climate parameter or possibly multiple climate parameters at once? 
     - For weights, I need to think of a good way to make sure we're calling the right weights file from the previous step and I'm not sure how to do that yet, but it will probably be linked to the naming convention used to save the weights file in step 1.  
- What are the calculations after the keyed merge? Is it multiplying the climate value by the geoweight and then summing across the polygon? So something like:
     - weighted_value = climate value * geoweight 
     - sum_value = sum(weighted value) by poly id 
- What's the ideal output of this step? A single weighted value of the climate parameter by polygon (by year?), or do we still want to have the raster gridnumbers in there as well?

**Comment by tcarleton** [comment]
All good questions @saraorofino ! 

1. I think you want to be able to call this function and pass it one climate variable you'd like. I suppose we could pass a vector of climate variables but that gets confusing if you want different nonlinearities for different variables or if there are other things that differ across variables. So I'd probably say an argument of the function is which climate variable you want to run and then the function just performs the nonlinear transformation and the aggregation for that one variable for the years specified. Weights will depend on a) climate data source; b) weighting scheme; c) polygons. So we need a convention for naming weights files based on those arguments.
2. Calculations are: i) nonlinear transformation, if any, of the climate data at the grid-by-day level; ii) weighted average of the output from step i) using the weights from the previous stage of the analysis. To start, can you code up options as simple average (e.g. average daily temperature, no nonlinearities), polynomials of power p (e.g, average daily temperature squared), and bins (e.g. number of days that fall within a set of 1 degree bins). Can start simple and add on here as we go forward. 
3. Ideal output is monthly sums of the output from step 2 for each polygon. We don't need to save the raster grid numbers anymore.

**Comment by saraorofino** [comment]
@tcarleton  @anna-boser  Sadly, the lazy raster is not working as well as I had hoped. I changed the lon in the weights table to be 0-360 to match the lon of the raster and then cropping/loading works but there are a few issues: 

- The weights extent (`weights_ext` object) is x: 144.5 to 295.5 and y: -14.5 to 71.5; the original resolution of the climate raster is 0.25 x 0.25. But when I crop the lazy raster to this extent and then read it into memory as a raster two things go wrong: 
   - 1) the resolution is now 0.1135522 x 0.2063397 which is probably leading to the second problem
   - 2) the extent is 144.375 to 295.625 and y: -14.625 to 71.625 
   - It also has no crs and it doesn't show a crs when I check the lazy raster object even though I think era5 should have a crs since I matched the polygons to the raster crs in Step 1 

This is the code that makes the lazy raster and then converts to an in memory raster and I just copied it straight from the `lazyraster` package documentation.

`clim_lazy <- raster::crop(lazyraster(nc_file), weights_ext)`     
`clim_raster <- lazyraster::as_raster(clim_lazy)` 

I'm including a screenshot of the comparison between the lazy raster and the in memory climate raster. 

<img width="763" alt="Screen Shot 2022-03-03 at 3 12 53 PM" src="https://user-images.githubusercontent.com/43861300/156668683-16482fc5-86d2-43e3-b2af-b4cb0865876e.png">
 
I'm also confused about the time aspect of the raster. The "window dimension" in the `clim_lazy` object was what I was expecting to be the ncol, nrow of my climate raster but it isn't. I thought maybe it tried to read it in as one flat matrix but nrow x ncol x 24 is still not the same as the number of cells in the raster. I tried to read in `clim_lazy` as a raster stack but I'm getting errors about the differing number of rows. 

The other problem stemming from this is that the `gridNumber` of the geoweights, which is what I was planning to merge on, was created using 1:length(clim_raster) in step 1 which was using an entire world raster. Making a data table of this new cropped climate raster using 1:length(clim_raster) isn't going to give me grid numbers that actually line up. I thought I could use xy values since they're in the geoweights table now and I found a function someone made to turn rasters in to a data.table and include the xy values. I think this will work once I figure out the resolution problem, but right now it doesn't work because the xy values are at different resolutions so they don't match up. 

I'm wondering if we need to abandon lazy loading even though that will definitely slow things down. I think the `lazyraster` package is pretty new so I'm not sure we'll be able to find a lot of issues or online help with figuring out why it's not lining up the way we think it should. But also I could just be missing something.
